src/transformers/cache_utils.py

```diff
@@ -292,6 +294,21 @@ def update(
         if layer_idx == 0:
             self._seen_tokens += key_states.shape[-2]
 
+        # Update the sin/cos cache, which holds sin/cos values for all possible positions
+        if using_rope and layer_idx == 0:
+            # BC: some models still pass `sin`/`cos` with 2 dims. In those models, they are the full sin/cos. Remove
+            # after all RoPE models have a llama-like cache utilization.
+            if cos.dim() == 2:
+                self._cos_cache = cos
+                self._sin_cache = sin
+            else:
+                if self._cos_cache is None:
+                    self._cos_cache = cos[0, ...]
+                    self._sin_cache = sin[0, ...]
+                elif self._cos_cache.shape[0] < self.window_length:
+                    self._cos_cache = torch.cat([self._cos_cache, cos[0, ...]], dim=0)
+                    self._sin_cache = torch.cat([self._sin_cache, sin[0, ...]], dim=0)
```

Comment:
Just for my own understanding of how the cache is meant to work, I have two Qs: 

1. Values passed in on `update` call
if we call `update` with `sin` and `cos` passed in, is the cache keeping old values + new values i.e. `self._cos_cache[:self._cos_cache_prev.shape[0]]` are the old values and `self._cos_cache[self._cos_cache_prev.shape[0]:]` is the new values, or the passed in `cos` is just the new values to be appended? 

2. Window length
Is the assumption here that the window length is constant once the cache is created?

Reply:
@amyeroberts 

1. The values passed in `cos` are new values to be appended. In RoPE models, `sin` and `cos` are a constant with shape `[config.max_position_embeddings, rope_embedding_dims, config.hidden_size // config.num_attention_heads]`. However, with the compile-optimized modeling code, we only materialize the needed parts of this matrices, with shape[0] = `input_ids.shape[1]` = input sequence length. Since `SinkCache` needs access to all `sin` and `cos` values up to shape[0] = `self.window_length` when going beyond the window length, this cache was created. 

Alternatively, we could pass the the model config to compute the full `sin` and `cos`, but that would be (IMO) an ugly interface (we would have to use the model config to instantiate a RoPE layer inside the cache, to then compute these values and discard the layer).

______________________

2. Yes. `SinkCache` is a fixed-length cache -- its purpose is to be used with `self.window_length` < `config.max_position_embeddings`, while enabling coherent outputs beyond full sequence length = `self.window_length`. In other words, coherent long outputs with a relatively short cache :) Its limitation is that it can only recall content back up to the size of the window length, it quickly forgets things.

Reply:
Got it - thanks for taking the time to write this up and explain!